### PR TITLE
Add the relaxed NAF

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

This PR adds the relaxed NAF, which differs from the original `find_wnaf` (should be `find_naf`?) by allowing the top 2 bits to be both nonzeros. This can, in some cases, reduce the overhead.

The algorithm is indeed quite simple. The only case when the standard NAF may be suboptimal (note: NAF is optimal for hamming weight, but not the length + the hamming weight) is when the NAF ends with "-1 0 1", which can be systematically changed into "1 1". There are no other cases.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by Pratyush** [comment]
Is this used anywhere? (Eg: for faster scalar muls?)

File: ff/src/biginteger/arithmetic.rs [hunk 1]
```diff
@@ -104,3 +104,62 @@ pub fn find_wnaf(num: &[u64]) -> Vec<i64> {
 
     res
 }
+
+pub fn find_relaxed_naf(num: &[u64]) -> Vec<i64> {
```
Comment (ValarDragon) [inline ff/src/biginteger/arithmetic.rs:108]: Can the relaxed naf definition be commented here

**Comment by weikengchen** [comment]
> Is this used anywhere? (Eg: for faster scalar muls?)

I originally planned to use it for Poseidon. Unfortunately, for essentially all the prime numbers being considered, NAF is not better than the original form, because when NAF is better, it needs to use "inverse", and "inverse" here is costly, which requires 3 constraints to compute (it is needed to be compatible with the case that `d` is zero).

Maybe it would be more useful for EC scalar mul, in which "inverse" (aka negative) is cheap.

File: ff/src/biginteger/arithmetic.rs (hunk 1 above)
Comment (weikengchen) [reply ff/src/biginteger/arithmetic.rs:108]: Yes, let me add some description to it.

**Comment by weikengchen** [comment]
On a separate note, should we rename `find_wnaf` to `find_naf`? there is no window.

**Comment by Pratyush** [comment]
Yeah, `find_naf` makes sense.

**Comment by Pratyush** [comment]
@weikengchen what's the status of this PR?

**Comment by weikengchen** [comment]
I should be able to revisit this PR today or tomorrow.

File: ff/src/biginteger/arithmetic.rs (hunk 1 above)
Comment (weikengchen) [reply ff/src/biginteger/arithmetic.rs:136]: Added.

**Comment by weikengchen** [comment]
The PR is ready to go. @Pratyush please take a look

File: ff/src/biginteger/arithmetic.rs [hunk 2]
```diff
@@ -119,3 +119,75 @@ pub fn find_wnaf(num: &[u64]) -> Vec<i64> {
...
+        res.resize(len - 1, 0);
+    }
+
+    res
+}
+
+#[test]
+fn test_find_relaxed_naf_usefulness() {
+    let vec = find_relaxed_naf(&[12u64]);
+    assert_eq!(vec.len(), 4);
```
Comment (Pratyush) [inline ff/src/biginteger/arithmetic.rs:152]: Maybe we can add here `find_naf(&12u64])` which has a shorter length?

**Review by Pratyush — APPROVED** [review]

File: ff/src/biginteger/arithmetic.rs (hunk 2 above)
Comment (weikengchen) [reply ff/src/biginteger/arithmetic.rs:155]: Yes. Added. It has a slightly longer len---5.

**Pull request merged**